src/Eval.hs

```diff
@@ -847,21 +847,47 @@ specialCommandSet ctx [x@(XObj (Sym path@(SymPath mod n) _) _ _), value] = do
   case result of
     Left err -> return (newCtx, Left err)
     Right evald -> do
-      let globalEnv = contextGlobalEnv ctx
-          elem = XObj (Lst  [XObj DefDynamic Nothing Nothing, XObj (Sym path Symbol) Nothing Nothing, evald]) (info value) (Just DynamicTy)
-          binder = (Binder emptyMeta elem)
-          nctx = newCtx { contextGlobalEnv=envInsertAt globalEnv path binder }
-      case contextInternalEnv nctx of
-        Nothing -> return (nctx, dynamicNil)
-        Just env ->
-          if contextPath nctx == mod
-          then return (nctx{contextInternalEnv=Just (envReplaceBinding (SymPath [] n) binder env)}, dynamicNil)
-          else return (nctx, dynamicNil)
+      let typeEnv = contextTypeEnv ctx
+          globalEnv = contextGlobalEnv ctx
+          (newEnv, originalType) = setStaticOrDynamicVar path evald globalEnv
+      -- Annotate to determine the type of the evaluated value.
+      (nctx, r) <- annotateWithinContext False newCtx evald
+      case r of
+        Left err -> return (ctx, Left err) -- type indeterminable
+        Right (annXObj@(XObj _ i t), deps) ->
+          if originalType /= Just DynamicTy && (t /= originalType)
+          then return (evalError ctx ("can't `set!` " ++ show path ++ " to a value of type " ++ show (fromJust t) ++ ", " ++ show path ++ " has type " ++ show (fromJust originalType)) (info x))
+          else case contextInternalEnv nctx of
+                 Nothing ->
+                   -- The variable is global; update the global env; redefine
+                      return (nctx{contextGlobalEnv = newEnv}, dynamicNil)
+                 Just env ->
```

Comment:
Yep, you are correct, good catch! In the current implementation, `x` would only be temporarily set to `y` in the `let` scope:

```clojure
(def x 2)
(let [y 3] (set! x y))
(Int.+ 1 x )
=> 3
(def x 2)
(let [y 3] (do (set! x y) (Int.+ 1 x)))
=> 4
```

Which definitely isn't the intention--Carp probably wants to permit such manipulation of global variables for game dev! Lemme look into a fix.

As a side note, it's also presents an interesting language design choice, I think. I find the notion of a language that only permits side effects in the direct parent scope (and not grandparents or children) kind of interesting. 

I think one thing that is kind of unfortunate about Carp is that the type system currently won't communicate that a let like your example is side effecting and manipulates the global scope if one uses `do`!

```clojure
(defn bar [] (let [y 3] (do (set! x y) (Int.+ 1 x))))
:t bar
=> bar : (Fn [] Int)
``` 

The side-effects are hidden! Of course, if the side-effect is the only thing the function does, it's typed correctly:

```clojure
(defn foo [] (let [y 3] (set! x y)))
:i foo
=> foo : (Fn [] ())
```

I personally find the `do` case a bit unsettling if we allow variables outside the current scope to be altered in other scopes. I think if we do allow it, we should also introduce a type for `T with side effects`.

Reply:
Or, we make Carp more "pure" and disallow manipulation of variables in other scopes--in which case something like that `let` block should issue an error about `x` being outside of the current scope (I think we already do this for interior/child scopes within functions?)

However, that would definitely have impacts on certain use-cases like game-dev, since it would make the manipulation of global state more difficult--or I suppose actually impossible in the case of top-level declarations (since the manipulation would presumably happen in `main` or another lower scope).

Reply:
I kind of like the idea of allowing the change of global state and introducing some new type annotation that represents "functions with side effects that return a value"

This would make it easy to use carp for domains that benefit from manipulation of global states (like games) while also making it possible to enforce what areas of the code should be "purer" than others.

e.g. Assuming this `!` as a temporary notation for the a function with side-effects: `(Fn [] Int)!`, one could do the following:

```clojure
(definterface map (Fn [(Fn [a] b) (f a)] (f b))) ;; to define map that *only* allows functions free of side-effects
(definterface effect-map (Fn [(Fn [a] b)! (f a)] (f b))) ;; to define map that allows functions that possibly perform a side-effect
```

One could then infer that any function that relies on a function with a side-effect in turn potentially has side-effects.

Or we could use a type constructor for modeling side-effects in the type system:

```
(Fn [(Effect (Fn [a] b))]) ;; of course, `Effect` would only take `Fn` types as inhabitants.
```

Then one could define, e.g. a function from a side-effecting function to its "pure" variant:

```
(Effect (Fn [a] b))) -> (Fn [a] b)
```

A nice way of hinting at the relationship between e.g. a function that performs repetitions via iteration and one that performs them via recursion, taking an iterative function as a recursive function that performs repetition through the side-effecting manipulation of state, rather than direct recursion.

In either case, the type would be somewhat magical, since in order to enforce it requires the compiler assigning a type based on the presence (or absence) of `do`.

Reply:
While I think this is an interesting idea, I believe that that should be a separate issue/PR. Do you agree?

Reply:
I'll open a separate issue, thanks!

Reply:
Did you end up opening an issue for this?

Reply:
(Sorry for the thread necromancy)